src/ray/gcs/store_client/in_memory_store_client.h

```diff
@@ -73,25 +74,26 @@ class InMemoryStoreClient : public StoreClient {
  private:
   struct InMemoryTable {
     /// Mutex to protect the records_ field and the index_keys_ field.
-    absl::Mutex mutex_;
+    mutable absl::Mutex mutex_;
     // Mapping from key to data.
+    // TODO(dayshah): benchmark reader/writer locks against boost::concurrent_flat_map
     absl::flat_hash_map<std::string, std::string> records_ ABSL_GUARDED_BY(mutex_);
   };
 
-  std::shared_ptr<InMemoryStoreClient::InMemoryTable> GetOrCreateTable(
-      const std::string &table_name) ABSL_LOCKS_EXCLUDED(mutex_);
+  InMemoryTable &CreateOrGetMutableTable(const std::string &table_name);
+
+  std::optional<const InMemoryTable *> GetTable(const std::string &table_name);
```

Comment:
Could you please add a comment on (1) not found case; (2) memory ownership (how we make sure returned pointer is valid).

Reply:
Seems every time you deference twice, what about returning pointer directly?

Reply:
also I think we can simply return `const InMemoryTable *`? Do we need a std::optional ?

Reply:
ya, was trying to go with using pointer boxed in optional for safety, but updated to return nullptr and have caller check for it, more consistent with the rest of the codebase too